### PR TITLE
Fixes a bug in complex_promotion(x, Type) and in complex_promotion(x,…

### PR DESCRIPTION
**Pull request by camillew**
… y),

which together were still resulting in the desired behavior. Specifically, complex_promotion(x, Type) was a no-op when x was already a complex (even if not with type Type). We are changing this so that x is converted to the other complex representation if needed. Conversely, we are changing the complex_promotion(x, y) implementation, so that it does not force x and y to have the same complex representation when they are both already complex numbers. This is because we should defer to the actual op evaluation to decide how to best handle them the mix of argument representations.

In other words, we fixed complex_promotion(x, Type) for the case where x is already a complex, but we also updated complex_promotion(x, y) so that it does not rely on that case anymore.

### CODE REVIEW

**Review by c3d — APPROVED** [review]
Hi @camillew.

The fix looks correct.  I will run through the tests to see if the tests need adjustments.

**Comment by c3d** [comment]
Merged in `dev`.
